libs/jsonschema/from_type.go

```diff
@@ -0,0 +1,335 @@
+// in the schema.
+func FromType(typ reflect.Type, fns []func(typ reflect.Type, s Schema) Schema) (Schema, error) {
+	c := constructor{
+		definitions: make(map[string]Schema),
+		seen:        make(map[string]reflect.Type),
+		root:        typ,
+	}
+
+	err := c.walk(typ)
+	if err != nil {
+		return Schema{}, err
+	}
+
+	for _, fn := range fns {
+		for k, v := range c.definitions {
+			c.definitions[k] = fn(c.seen[k], v)
+		}
+	}
+
+	res := c.definitions[typePath(typ)]
+	res.Definitions = c.Definitions()
+	return res, nil
+}
+
+// typePath computes a unique string representation of the type. $ref in the generated
+// JSON schema will refer to this path. See TestTypePath for examples outputs.
+func typePath(typ reflect.Type) string {
+	// Pointers have a typ.Name() of "". Dereference them to get the underlying type.
+	for typ.Kind() == reflect.Ptr {
+		typ = typ.Elem()
+	}
+
+	if typ.Kind() == reflect.Interface {
+		return "interface"
+	}
+
+	// Recursively call typePath, to handle slices of slices / maps.
+	if typ.Kind() == reflect.Slice {
+		return path.Join("slice", typePath(typ.Elem()))
+	}
+
+	if typ.Kind() == reflect.Map {
+		if typ.Key().Kind() != reflect.String {
+			panic(fmt.Sprintf("found map with non-string key: %v", typ.Key()))
+		}
+
+		// Recursively call typePath, to handle maps of maps / slices.
+		return path.Join("map", typePath(typ.Elem()))
+	}
+
+	switch {
+	case typ.PkgPath() != "" && typ.Name() != "":
+		return typ.PkgPath() + "." + typ.Name()
+	case typ.Name() != "":
+		return typ.Name()
+	default:
+		// Invariant. This function should return a non-empty string
+		// for all types.
+		panic("unexpected empty type name for type: " + typ.String())
+	}
+}
+
+// Walk the Go type, generating $defs for every type encountered, and populating
+// the corresponding $ref in the JSON schema.
+func (c *constructor) walk(typ reflect.Type) error {
+	// Dereference pointers if necessary.
+	for typ.Kind() == reflect.Ptr {
+		typ = typ.Elem()
+	}
+
+	typPath := typePath(typ)
+
+	// Return early if the type has already been seen, to avoid infinite recursion.
+	if _, ok := c.seen[typPath]; ok {
+		return nil
+	}
+	c.seen[typPath] = typ
+
+	var s Schema
+	var err error
+
+	switch typ.Kind() {
+	case reflect.Struct:
+		s, err = c.fromTypeStruct(typ)
+	case reflect.Slice:
+		s, err = c.fromTypeSlice(typ)
+	case reflect.Map:
+		s, err = c.fromTypeMap(typ)
+	case reflect.String:
+		s = Schema{Type: StringType}
+	case reflect.Bool:
+		s = Schema{Type: BooleanType}
+	case reflect.Int, reflect.Int32, reflect.Int64:
+		s = Schema{Type: IntegerType}
+	case reflect.Float32, reflect.Float64:
+		s = Schema{Type: NumberType}
+	case reflect.Interface:
+		// Interface or any types are not serialized to JSON by the default JSON
+		// unmarshaller (json.Unmarshal). They likely thus are parsed by the
+		// dynamic configuration tree and we should support arbitary values here.
+		// Eg: variables.default can be anything.
```

Comment:
Yeah my bad. Simplified the comment. What I meant was that for interface types the Go type system does not provide any relevant information. And that we typically process this using dyn library rather than the default `json.Unmarshaller`.